### ISSUE TITLE
[Epic] v3.0: Surrogate Model Training & ONNX Export

### ISSUE THREAD

**Issue by anchapin** [anchapin/fluxion#718]
## Overview

Train compact ML surrogate models per physics component and export to ONNX for Rust-native inference via the existing `ort` integration.

**Prerequisite**: Phase 3 (v2.1) synthetic data generation must be complete.

## Architecture

The existing `src/ai/` module already provides:
- `SurrogateManager` with full ONNX Runtime integration (CPU/CUDA/CoreML/DirectML/OpenVINO)
- `ComponentSurrogate` + `CompositeSurrogate` for modular inference
- Quantization support (FP32/FP16/INT8)
- Domain validation via `SurrogateDomain`
- `NeuralWithFallback` mode (currently falls back to analytical — **this is what we're filling in**)

The Rust inference layer is ready. This epic builds the **trained models** that plug into it.

## Candidate Model Architectures

| Architecture | Pros | Cons |
|---|---|---|
| **MLP (baseline)** | Fast, easy ONNX export, interpretable | May underfit complex interactions |
| **Gaussian Process** | Uncertainty quantification, good for small data | Doesn't scale to 1M+ samples |
| **XGBoost/LightGBM** | Excellent tabular performance, ONNX via `sklearn-onnx` | Larger model size |
| **Ensemble (MLP+GP)** | Best of both — accuracy + uncertainty | More complex training/export |

## Target Metrics
- **RMSE < 2%** vs physics baseline on held-out test set
- **Inference time < 1ms** per prediction (measured in Rust via `ort`)
- **ONNX opset 17+** compatible with `ort` 2.0.0-rc.10

## Training Stack
- Python 3.12 + PyTorch or scikit-learn
- `torch.onnx.export` or `skl2onnx` for ONNX conversion
- `onnxruntime` for Python-side validation pre-export
- Output: `models/surrogate_{component}.onnx`

## Deliverables
- [ ] `scripts/train_surrogate.py` — training script per component
- [ ] `scripts/export_onnx.py` — ONNX export + validation
- [ ] `scripts/validate_surrogate.py` — accuracy benchmarking vs physics
- [ ] `models/surrogate_{component}.onnx` — trained model files
- [ ] `docs/surrogate_model_card.md` — architecture, metrics, training details
- [ ] Integration test: Rust `SurrogateManager::new()` loads model + passes accuracy check

## Acceptance Criteria
- [ ] RMSE < 2% for each target component
- [ ] `cargo test` passes with surrogate models present
- [ ] `SurrogateMode::NeuralWithFallback` uses neural path (not analytical fallback)
- [ ] ONNX models load cleanly via `ort` in both CPU and CUDA builds

## Related
- Depends on: v2.1 Synthetic Data Generation
- Related research: #708 (MIRAI temporal forecasting — outputs from this epic feed that work)
- Feeds into: v3.1 Modular Hybrid Architecture

**Comment by anchapin** [comment]
## Architecture Decision (2026-05-12)

**MLP-first, MLP+GP ensemble as follow-on.**

### Phase 4a: MLP Baseline (primary deliverable)
- Single hidden-layer MLP per physics component
- Fast, auditable, easy ONNX export via `torch.onnx.export` (opset 17, `ort` 2.0.0-rc.10 compatible)
- Target: RMSE < 2%, inference < 1ms in Rust

### Phase 4b: MLP+GP Ensemble (follow-on, unlocks #708)
- GP layer on MLP embeddings for uncertainty quantification
- Enables probabilistic multi-horizon forecasting needed for MIRAI agentic orchestration (#708)
- Will be tracked in a sub-issue once Phase 4a is validated

**Comment by anchapin** [comment]
## Literature Scan: PINNs for Building Thermal Simulation & ONNX/Rust Pipeline

*Research & Innovation Engineer — May 12, 2026*

Coordinating with ML & Surrogate Modeling Engineer (see #764 for MLP spec). This comment addresses the PINN literature scan and ONNX pipeline; synthetic data findings are on #719.

---

## Key Question: CTF Outputs → Pure MLP vs. PINN?

> Are fluxion's CTF solver outputs suitable training targets for a pure MLP, or would a physics-informed loss term significantly reduce data volume?

**Answer: Both paths work with CTF outputs, but a PINN physics-informed loss reduces required training samples by ~10–30× while improving physical consistency at long horizons.**

| Approach | Typical training samples | Physical consistency | Notes |
|----------|--------------------------|---------------------|-------|
| Pure MLP | 50K–200K | Not guaranteed | Higher volume, fastest inference |
| PINN (physics-informed loss) | **1,766–15,000** | Enforced via ODE residual | IBPSA 2025 / ScienceDirect |
| RC model (analytical) | N/A | By construction | Baseline only |

CTF outputs are well-suited as PINN targets because they satisfy the RC energy balance by construction — the physics residual will naturally converge to CTF-consistent predictions, making λ_physics calibration straightforward.

**Recommended phasing:**
- **Phase 4a** (per Alex's priority): Pure MLP baseline — CTF outputs as training targets, 50K samples, no architecture changes needed
- **Phase 4b**: Add 5R1C ODE physics loss term to same MLP → reduces training set to 5K–15K, improves extrapolation

> **Note**: PINN trained on CTF outputs will inherit CTF errors for high-mass (900-series) constructions. For Phase 4b, consider training on FD solver outputs (pending #726).

---

## Relevant 2024-2025 Papers

### 1. IBPSA 2025 — Inverse PINN for Building Thermal Parameter Identification
- Inverse PINN identifies R/C values from smart thermostat data across **234 Illinois homes**
- Architecture: MLP + 2R2C ODE residual; Optuna hyperparameter tuning
- Test MSE: 0.81; demonstrates **cross-building generalization**
- Key insight: λ_physics weight sensitivity analysis is essential before deployment (optimal ~0.1–0.3)

### 2. ScienceDirect — PINN for Cooling Electricity Prediction (Historic Buildings)
- PINN achieves accurate predictions with **only 1,766 EnergyPlus training samples**
- Two-stage prediction: T_operative → E_cooling
- Pure MLP fails at this data volume; PINN maintains physical consistency

### 3. Academia.edu — PINN on ASHRAE RP-1312 / BOPTEST Benchmarks
- Systematic comparison: PINN vs. RC grey-box vs. pure NN on standardized BEM benchmarks
- PINN **outperforms both on long-horizon forecasting** while pure NN has fastest inference
- Directly applicable to fluxion's ASHRAE 140 validation target

### 4. MPC Application (2024)
- PINN-based MPC improves thermal comfort by **14.6–31.9%**, reduces energy cost **18.3–29.3%**
- Confirms PINN accuracy is sufficient for control-quality surrogate targets

### 5. RC Network Integration — 2024 Review
- 2R2C PINN enhances indoor temperature prediction accuracy by up to **40% vs. pure RC**
- First-principles ODE constraints prevent thermodynamically impossible extrapolations

---

## ONNX Export → Rust Inference Pipeline

### Recommended Stack

| Layer | Library | Notes |
|-------|---------|-------|
| Training | PyTorch | `torch.onnx.export(dynamo=True)` |
| ONNX validation | Netron | Verify input/output names before Rust integration |
| Rust inference (primary) | **`ort` v2.0+** | Best perf; CUDA support; 49MB vs PyTorch 1.6GB |
| Rust inference (fallback) | **`tract-onnx` v0.21+** | Pure Rust, no C++ deps, WASM-compatible, CPU-only |

### Export Pattern
```python
example_input = torch.randn(1, n_features)  # your BEM input dim
onnx_program = torch.onnx.export(model, example_input, dynamo=True)
onnx_program.save("fluxion_surrogate.onnx")
```

### Rust Inference (ort)
```toml
[dependencies]
ort = "2.0"
ndarray = "0.15"
```
```rust
let session = Session::builder()?
    .with_optimization_level(GraphOptimizationLevel::Level3)?
    .with_intra_threads(4)?
    .commit_from_file("fluxion_surrogate.onnx")?;
let outputs = session.run(ort::inputs!["input" => tensor])?;
```

### Notes for BatchOracle
- Use `dynamic_axes` in export so `BatchOracle` can vary batch size at runtime
- For rayon parallel hot path: `ort` with `Level3` optimization
- For WASM/edge targets: `tract` eliminates C++ runtime dependency

---

Full scan document (with #719 synthetic data and LHC bounds): see companion comment on #719 and report linked from #764.

**Comment by anchapin** [comment]
**FROZEN — Bottom-Up Validation Restructuring**

This issue is frozen as part of the architectural pivot from top-down ASHRAE 140 calibration to bottom-up physics validation.

**Why frozen:** We are no longer tuning/calibrating the engine against ASHRAE 140 results. Instead, we are:
1. Building ARCHITECTURE.md with strict module boundaries
2. Generating EnergyPlus reference data for isolated physics modules
3. Unit-testing each module (Solar, Conduction, Infiltration, Zone Balance) independently
4. Only resuming ASHRAE 140 validation AFTER individual modules pass 1% tolerance tests

**When this will be revisited:** After Phase 3 (Module Isolation & Data-Driven Testing) completes.

See: Issue #[TBD - Architecture & Roadmap]

**Comment by anchapin** [comment]
## SPEC AMENDMENT — Modular architecture + HVAC mechanical minimum constraints (from #1178 + #1179)

---

### Amendment A: Modular MLP as Phase 4a candidate (from #1178)

Add the following row to the candidate architecture table:

| Architecture | Pros | Cons |
|---|---|---|
| **Modular MLP** (3 modules: Envelope / HVAC / Gains) | Per-subsystem physics constraints; independent retraining; mirrors fluxion solver decomposition; reduced LHC re-sweep on single-subsystem change | 3 ONNX models instead of 1; slightly more complex Rust integration via `CompositeSurrogate` |

**Priority alignment:** MLP-only first (user-confirmed preference preserved). Modular vs. monolithic is an architectural choice _within_ Phase 4a, not a new phase. Decision deferred to benchmark documented in #764 amendment.

**Phase 4b implication:** If GP ensemble is pursued, modular base models enable per-subsystem uncertainty quantification — envelope uncertainty ≠ HVAC uncertainty — which is a strict improvement over a single GP UQ layer on a monolithic MLP.

---

### Amendment B: HVAC surrogate must encode mechanical minimum power as hard constraints (from #1179)

**Source:** arXiv 2606.01665 — "Quantifying the Energy Floor: Direct Measurement and Replay Buffer Bias in SAC-Based HVAC Control"

**Finding:** The SAC energy floor in HVAC control is dominated by **equipment mechanical minimums** (99.8% continuous electrical loads). Best SAC achieves $35.57/day vs. measured floor of $35.51/day — a 0.17% gap. A 10K expansion of supply water temperature range yields only $0.03/day. The constraint is the equipment, not the algorithm.

**Implication for the HVAC surrogate:** If the surrogate models mechanical minimums as soft cost terms or ignores them entirely, it will systematically underestimate the true energy floor and produce training targets that suggest more headroom than physically exists. The RL policy will learn against a phantom optimization surface.

**Spec changes:**
- Add `min_heating_power_kw` and `min_cooling_power_kw` as **inputs** to the HVAC module (Module B in #764 modular spec, or as additional features in the monolithic MLP HVAC input group)
- Encode mechanical minimums as **hard output constraints** (clamp/relu on output layer), not loss function penalties
- **New acceptance criterion:** HVAC surrogate energy floor must match physics baseline to within 1% at minimum-load operating points

---

_No implementation action required yet — spec documentation only._
